File: ruby/lesson 10/unique_paths_recursive.rb
# https://leetcode.com/problems/unique-paths/

def unique_paths(m, n)
    dp = Array.new(m + 1) { Array.new(n + 1) }
    dp[1][1] = 1
    unique_paths_impl(m, n, dp)
end


def unique_paths_impl(row, col, dp)
    return 0 if row < 0 || col < 0
    return dp[row][col] unless dp[row][col].nil?

    dp[row][col] = unique_paths_impl(row - 1, col, dp) + unique_paths_impl(row, col - 1, dp)
end

File: javascript/lesson 10/uniquePathsRecursive.js
// https://leetcode.com/problems/unique-paths/

function uniquePaths(m, n) {
    const dp = Array.from(Array(m + 1), () => new Array(n + 1));
    dp[1][1] = 1;
    return uniquePathsImpl(m, n, dp);
}

function uniquePathsImpl(row, col, dp) {
    if (row < 0 || col < 0) return 0;
    if (dp[row][col] > 0) return dp[row][col];

    dp[row][col] = uniquePathsImpl(row - 1, col, dp) + uniquePathsImpl(row, col - 1, dp);
    return dp[row][col];
}
